### PR TITLE
Feature Additions, GUI Tweaks, Fixes

### LINKED ISSUE #965: [Request]: Can you add trigger bot or aimbot ?
### Problem

_No response_

### Solution

pls add trigger bot or aimbot

### Reason

make menu perfect

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Seanghost117**
**Nearby Option Feature Additions:**
[ Peds ]
- Delete nearby peds
- Kill nearby peds
- Kill nearby enemies (most critical important feature: https://www.unknowncheats.me/forum/3675913-post762.html)
- Peds Ignore
- Ped Rain (Just rains the nearby peds until they hit the ground and repeats)
- High Alert (Increase ped perception of nearby peds
- Ped Rush (sets the peds movement so that they move quickly

[ Vehicles ]
- Delete
- Max Upgrade
- Fully Downgrade

**Weapon Feature Additions**
[ Aimbot ]
- Aimbot enable / disable
- Fov change for aimbot fov
- Select Bone to lock on
- Smoothing + Speed
- Entity checks for (players, enemies, cops)
- Distance check

[ General ]
- Triggerbot

**Feature Tweaks**
[ Blackholes Changes ]
- Added a slider to change the radius of black holes
- Changed the toggles to command toggles

**Extra**
[ Misc ]
- Moved Primexz's get entity function to entities
- Extended buttons so that they can support sizing and color changes (maybe not the best way for colors but it works)

Suggest anything to add, please don't merge this in yet I want to hear some suggestions and fix up how the UI is, I'm going to make use of the sizing hack and make the buttons equal in sizing and spacing. Also mem leak?

### CODE REVIEW

**Review by Yimura — DISMISSED** [review]
Looks good to me for now.

File: src/backend/looped/world/nearby/ignore.cpp [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+#include "util/entity.hpp"
+
+namespace big
+{
+
+	class ignore : looped_command
+	{
+		using looped_command::looped_command;
+
+		virtual void on_tick() override
```
Comment (maybegreat48) [inline src/backend/looped/world/nearby/ignore.cpp:19]: Untested, but

on_enable
```
PLAYER::SET_EVERYONE_IGNORE_PLAYER(self::id, true);
PLAYER::SET_POLICE_IGNORE_PLAYER(self::id, true);
```

on_tick
```
EVENT::SUPPRESS_SHOCKING_EVENTS_NEXT_FRAME();
EVENT::SUPPRESS_AGITATION_EVENTS_NEXT_FRAME();
```

on_disable
```
PLAYER::SET_EVERYONE_IGNORE_PLAYER(self::id, false);
PLAYER::SET_POLICE_IGNORE_PLAYER(self::id, false);
```

File: src/backend/looped/world/nearby/peds_rain.cpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+	class ped_rain : looped_command 
+	{
+		using looped_command::looped_command;
+
+		virtual void on_tick() override
+		{
+
+			for (auto ped : entity::get_entities(false, true))
+			{
+				if (!ENTITY::IS_ENTITY_IN_AIR(ped)) { // Flag 17 = PED_FLAG_BLOCK_NON_TEMPORARY_EVENTS
```
Comment (maybegreat48) [inline src/backend/looped/world/nearby/peds_rain.cpp]: `PED_FLAG_BLOCK_NON_TEMPORARY_EVENTS` what?

Comment (Seanghost117) [reply src/backend/looped/world/nearby/peds_rain.cpp]: accidental comment I left, was in ignore.cpp

File: src/util/vehicle.hpp [hunk 3]
```diff
@@ -657,6 +657,17 @@ namespace big::vehicle
...
 	}
 
+	inline void downgrade(Vehicle vehicle)
+	{
+		VEHICLE::SET_VEHICLE_MOD_KIT(vehicle, 0);
+		for (int i = 0; i < 50; i++)
+		{
+			VEHICLE::REMOVE_VEHICLE_MOD(vehicle, i);
+		}
+		VEHICLE::REMOVE_VEHICLE_MOD(vehicle, 18); //Remove turbo on vehicle
```
Comment (maybegreat48) [inline src/util/vehicle.hpp]: Doesn't the above loop remove 18 and 22?

File: src/backend/looped/world/nearby/ignore.cpp (hunk 1 above)
Comment (maybegreat48) [reply src/backend/looped/world/nearby/ignore.cpp:19]: Can you test the `on_tick` part? Getting all peds for an ignore player feature seems too performance intensive IMO

File: src/backend/looped/world/nearby/high_alert.cpp [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+	class high_alert : looped_command
+	{
+		using looped_command::looped_command;
+
+		virtual void on_tick() override
+		{
+
+			for (auto ped : entity::get_entities(false, true))
+			{
```
Comment (maybegreat48) [inline src/backend/looped/world/nearby/high_alert.cpp:17]: Reset on `on_disable`. Same for the speed thing

Comment (Seanghost117) [reply src/backend/looped/world/nearby/high_alert.cpp:17]: right, thanks

**Comment by Seanghost117** [comment]
image
Tried to match the colors to the vibe

**Comment by Seanghost117** [comment]
**Update**
I changed the PR title as I've expanded this PR to more than just nearby options, I've added the following

[ Weapons ]
- Added aimbot
- Added fov, smoothing, and smoothing speed to aimbot options

[ GUI ]
- Added templating to components::button, seemed kind of weird to just throw it all in at the end, let me know if this is a problem @Yimura and I'll fix it

This should handle #965 , takes over part of #889 by @TheGreenBandit 
To Do:

- [x] Added ped relation check (for enemies, cops, etc)
- [x] Add Ped Bone selector for aimbot
- [ ] Update triggerbot to tap peds in vehicles as well
- [x] Include players
- [x] Distance check on the aimbot (and maybe triggerbot), limited by max range anyways

Also I am aware of slightly delayed / behind mouse tracking when it comes to moving entities at least at a specific speed and higher, I do plan to add some prediction in later but as this was just done I will handle this at a later time as it is not my priority + I'm tired

@maybegreat48 let me know if you have any suggestions or if I missed something in this commit

Preview (Just aimbot)

https://user-images.githubusercontent.com/36561158/218394645-ca12d6f7-5eda-4d56-b779-52ecfa63cb0e.mp4

File: src/backend/looped/weapons/triggerbot.cpp [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+	class triggerbot : looped_command
+	{
+		using looped_command::looped_command;
+
+		virtual void on_tick() override
+		{
+			Entity crosshair_catch;
+			if (PAD::IS_DISABLED_CONTROL_PRESSED(0, (int)ControllerInputs::INPUT_AIM))
+			{
+				if (entity::raycast(&crosshair_catch))
```
Comment (maybegreat48) [inline src/backend/looped/weapons/triggerbot.cpp:17]: Does `GET_ENTITY_PLAYER_IS_FREE_AIMING_AT` work? That native is no longer detected after the anticheat changes

Comment (Seanghost117) [reply src/backend/looped/weapons/triggerbot.cpp:17]: It actually did work, that’s what I had it set to before but opted for raycast for uniformity

**Comment by TheGreenBandit** [comment]
For the distance check just compare self::pos to the bone position using SYSTEM::VDIST

**Comment by Seanghost117** [comment]
> For the distance check just compare self::pos to the bone position using SYSTEM::VDIST

Distance and entities checks should be done, haven’t staged the commit yet, got to work on players later today if I have free time thought.

**Comment by Seanghost117** [comment]
Done with the aimbot, if there are any suggestions you can leave them but they will be done at a later time with the prediction, I'm going to add and edit some nearby features and switch it to review before I sit on it for too long

**Comment by BeztNA** [comment]
> Done with the aimbot, if there are any suggestions you can leave them but they will be done at a later time with the prediction, I'm going to add and edit some nearby features and switch it to review before I sit on it for too long

can u add heist editor sir ?

**Comment by Seanghost117** [comment]
> Done with the aimbot, if there are any suggestions you can leave them but they will be done at a later time with the prediction, I'm going to add and edit some nearby features and switch it to review before I sit on it for too long

Will probably just add options for editing the color of nearby vehicles, fucking with them, and whatever else. Delete vehicles is buggy so will remove it for now. Also, it seems ped ignore can be a good alternative for evading the law without actually editing your wanted level

> can u add heist editor sir ?

No, this will not be added, especially since it falls too close to recovery.

**Comment by BeztNA** [comment]
> > Done with the aimbot, if there are any suggestions you can leave them but they will be done at a later time with the prediction, I'm going to add and edit some nearby features and switch it to review before I sit on it for too long
> 
> Will probably just add options for editing the color of nearby vehicles, fucking with them, and whatever else. Delete vehicles is buggy so will remove it for now. Also, it seems ped ignore can be a good alternative for evading the law without actually editing your wanted level
> 
> > can u add heist editor sir ?
> 
> No, this will not be added, especially since it falls too close to recovery.

thanks for answer, but when do this pr release ?

**Comment by Seanghost117** [comment]
I will try to get this pushed within a few hours tops. I'm just going to step out to eat first

**Comment by BeztNA** [comment]
> I will try to get this pushed within a few hours tops. I'm just going to step out to eat first

tks u

**Comment by PrimeMK105** [comment]
> Done with the aimbot, if there are any suggestions you can leave them but they will be done at a later time with the prediction, I'm going to add and edit some nearby features and switch it to review before I sit on it for too long

Option to not shoot at friends/people who in your organization/motoclub?

File: src/backend/looped/weapons/aimbot.cpp [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+					else if (((relation == 4) || (relation == 5)) && g.weapons.aimbot.on_enemy) // relation 4 and 5 are for enemies
+						goto aimbot_handler;
+
+					else if(((type == 6 && !PED::IS_PED_MODEL(ped, rage::joaat("s_m_y_uscg_01"))) || type == 27 || // s_m_y_uscg_01 = us coast guard 1 (techniaclly military)
+						PED::IS_PED_MODEL(ped, rage::joaat("s_m_y_ranger_01")) || PED::IS_PED_MODEL(ped, rage::joaat("s_f_y_ranger_01"))) // ranger models
+						&& g.weapons.aimbot.on_police)
+						goto aimbot_handler;
+
+					else if (g.weapons.aimbot.on_npc && !PED::IS_PED_A_PLAYER(ped))
```
Comment (Yimura) [inline src/backend/looped/weapons/aimbot.cpp]: Could you redo this code without using `goto` and adding some curly brackets after the if statements.

File: src/backend/looped/world/nearby/vehicles_rain.cpp [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+namespace big
+{
+
+	class vehicle_rain : looped_command
+	{
+		using looped_command::looped_command;
+
+		virtual void on_tick() override
+		{
+
```
Comment (Yimura) [inline src/backend/looped/world/nearby/vehicles_rain.cpp]: Random newline.

File: src/backend/looped/world/nearby/vehicles_rain.cpp [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+						Vector3 my_location = ENTITY::GET_ENTITY_COORDS(self::ped, 1);
+						my_location.x = my_location.x + (rand() % 100 + (-50));
+						my_location.y = my_location.y + (rand() % 100 + (-50));
+						my_location.z = my_location.z + (rand() % 20 + 100);
+						ENTITY::SET_ENTITY_COORDS(vehicles, my_location.x, my_location.y, my_location.z, 0, 0, 0, 0);
+						ENTITY::SET_ENTITY_VELOCITY(vehicles, 0.0, 0.0, -1.0);
+					}
+				}
+			}
+		}
```
Comment (Yimura) [inline src/backend/looped/world/nearby/vehicles_rain.cpp]: Mixed curly brace placement, preferably have the opening curly bracket start on a newline.

File: src/core/globals.hpp [hunk 9]
```diff
@@ -615,6 +635,20 @@ namespace big
...
+				float smoothing_speed = 2.f;
+				bool on_player = true;
+				bool on_enemy = false;
+				bool on_police = false;
+				bool on_npc = false;
+				float fov = 90.f;
+				float distance = 200.f;
+				std::uint32_t selected_bone = 0x796E; // Default to head
+				NLOHMANN_DEFINE_TYPE_INTRUSIVE(aimbot, enable, smoothing, smoothing_speed, fov, selected_bone)
+			} aimbot{};
```
Comment (Yimura) [inline src/core/globals.hpp:643]: newline after struct definition

File: src/gui/components/components.hpp [hunk 10]
```diff
@@ -62,6 +61,24 @@ namespace big
...
 		}
 
+		template<ImVec2 size = ImVec2(0, 0), ImVec4 color = ImVec4(0.24f, 0.23f, 0.29f, 1.00f)>
+		static bool button(const std::string_view text) {
+			bool status = false;
+			ImGui::PushStyleColor(ImGuiCol_Button, color);
+			if (ImGui::Button(text.data(), size))
+				status = true;
+			else
+				status = false;
```
Comment (Yimura) [inline src/gui/components/components.hpp]:
```cpp
status = ImGui::Button(text.data(), size);
```

File: src/gui/components/components.hpp [hunk 11]
```diff
@@ -62,6 +61,24 @@ namespace big
...
+			ImGui::PopStyleColor(1);
+			return status;
+		}
+
+		template<ImVec2 size = ImVec2(0, 0), ImVec4 color = ImVec4(0.24f, 0.23f, 0.29f, 1.00f)>
+		static void button(const std::string_view text, std::function<void()> cb) {
+			if (button<size, color>(text)) {
+				g_fiber_pool->queue_job(cb);
+			}
+		}
```
Comment (Yimura) [inline src/gui/components/components.hpp:78]: Newline below function

File: src/views/self/view_weapons.cpp [hunk 12]
```diff
@@ -153,5 +153,30 @@ namespace big
 
 			break;
 		}
+
+		ImGui::Separator();
+		components::sub_title("Aim Assistance");
+		components::command_checkbox<"triggerbot">(); ImGui::SameLine();
+		components::command_checkbox<"aimbot">();
+
+		if (g.weapons.aimbot.enable) {
```
Comment (Yimura) [inline src/views/self/view_weapons.cpp]: curly bracket on newline

File: src/views/world/view_nearby.cpp [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+				}
+		}); ImGui::SameLine();
+
+		components::button<ImVec2(110, 0), ImVec4(0.4549f, 0.03529f, 0.03529f, 1.f)>
+			("Downgrade", [] {
+			for (auto vehs : entity::get_entities(true, false))
+				if (entity::take_control_of(vehs)) {
+					vehicle::downgrade(vehs);
+					script::get_current()->yield();
+				}
```
Comment (Yimura) [inline src/views/world/view_nearby.cpp]: Wrap if between curly brackets and also opening curly bracket on newline

File: src/views/world/view_nearby.cpp [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+				for (auto peds : entity::get_entities(false, true)) entity::delete_entity(peds); 
+			}); ImGui::SameLine();
+		components::button<ImVec2(110, 0), ImVec4(0.70196f, 0.3333f, 0.00392f, 1.f)>
+			("Kill", [] { 
+				for (auto peds : entity::get_entities(false, true)) ped::kill_ped(peds); 
+			}); ImGui::SameLine();
+		components::button<ImVec2(110, 0), ImVec4(0.76078f, 0.f, 0.03529f, 1.f)>
+			("Kill Enemies", [] { 
+				for (auto peds : entity::get_entities(false, true)) ped::kill_ped_by_relation(peds, 4 || 5); 
+			});
```
Comment (Yimura) [inline src/views/world/view_nearby.cpp]: Not very readable though I'm open for suggestions.

```cpp
components::button<ImVec2(110, 0)> ("Delete", []
{ 
	for (auto peds : entity::get_entities(false, true))
		entity::delete_entity(peds); 
}); ImGui::SameLine();
components::button<ImVec2(110, 0), ImVec4(0.70196f, 0.3333f, 0.00392f, 1.f)>("Kill", []
{ 
	for (auto peds : entity::get_entities(false, true))
		ped::kill_ped(peds); 
}); ImGui::SameLine();
components::button<ImVec2(110, 0), ImVec4(0.76078f, 0.f, 0.03529f, 1.f)>("Kill Enemies", []
{ 
	for (auto peds : entity::get_entities(false, true))
		ped::kill_ped_by_relation(peds, 4 || 5); 
});
```

Same for "Max Upgrade" & "Downgrade" options below

**Review by Yimura — CHANGES_REQUESTED** [review]
I messed up merging the code and broke the build, I'll fix that when I get home.

File: src/backend/looped/world/nearby/vehicles_rain.cpp (hunk 7 above)
Comment (Seanghost117) [reply src/backend/looped/world/nearby/vehicles_rain.cpp]: fixed

File: src/backend/looped/weapons/aimbot.cpp (hunk 6 above)
Comment (Yimura) [reply src/backend/looped/weapons/aimbot.cpp]: I'm fine with just adding curly braces if the goto's make the code flow easier.

**Comment by BeztNA** [comment]
when do this pr release sir ?

**Comment by TheGreenBandit** [comment]
> when do this pr release sir ?

When it is considered polished and ready by sean

**Review by Yimura — APPROVED** [review]
Thanks a lot for adding so many new features, waiting for CI to finish and then I'll merge it.

**Pull request merged**